leptonai/api/v1/api_resource.py

```diff
@@ -0,0 +1,149 @@
+        self._delete = _client._delete
+        self._head = _client._head
+
+    # A type variable to represent a subclass of BaseModel
+    T = TypeVar("T", bound=BaseModel)
+
+    def safe_json(
+        self, content: Union[T, List[T]]
+    ) -> Union[Dict[str, Any], List[Dict[str, Any]]]:
+        """
+        A utility function to safely convert BaseModel or a list of BaseModel to
+        JSON serializable dictionary or list of dictionary. This also honors the alias
+        defined in the BaseModel.
+
+        Args:
+            content (Union[T, List[T]]): BaseModel or List[BaseModel]
+        Returns:
+            Union[Dict[str, Any], List[Dict[str, Any]]]: JSON serializable dictionary
+            or list of dictionary
+        Raises:
+            ValueError: If the input is not BaseModel or List[BaseModel]
+        """
+        if isinstance(content, BaseModel):
+            return content.dict(exclude_none=True, by_alias=True)
+        elif isinstance(content, list) and all(
+            isinstance(c, BaseModel) for c in content
+        ):
+            return [c.dict(exclude_none=True, by_alias=True) for c in content]
+        else:
+            raise ValueError(
+                "safe_json only accepts BaseModel or List[BaseModel] as input."
+            )
+
+    def _raise_if_not_ok(self, response: Response):
+        """
+        Raise a RuntimeError if the response is not ok. Given the
+        """
+        if response.status_code >= 400 and response.status_code < 500:
+            raise ClientError(response)
+        elif response.status_code >= 500:
+            raise ServerError(response)
+        return response
+
+    def _print_programming_error(self, response: Response, e: Exception) -> NoReturn:
+        """
+        Print a programming error message. This should not happen in production.
+        """
+        raise RuntimeError(
+            "You encountered a programming error. Please report this, and include"
+            " the following debug info:\n*** begin of debug info ***\nresponse"
+            " returned 200 OK, but the content cannot be decoded as"
+            f" json.\nresponse.text: {response.text}\n\nexception"
+            f" details:\n{e}\n*** end of debug info ***"
+        )
+
+    T = TypeVar("T", bound=BaseModel)
+
+    def ensure_type(self, response, EnsuredType: Type[T]) -> T:
+        """
+        Utility function to ensure that the response is of the given type.
+        """
+        self._raise_if_not_ok(response)
+        try:
+            return EnsuredType(**response.json())
+        except Exception as e:
+            self._print_programming_error(response, e)
+
+    def ensure_list(self, response, EnsuredType: Type[T]) -> List[T]:
+        """
+        Utility function to ensure that the response is a list of the given type.
+        """
+        self._raise_if_not_ok(response)
+        try:
+            return [EnsuredType(**item) for item in response.json()]
```

Comment:
nit: check `response.json()` is of list type first?

Reply:
Ah I basially forgo this because in any case, the action is to print out a programming error, so we don't need to check separately.